### ISSUE TITLE
Clang does not report invalid arithmetic operations at compile time when doing `_Complex` math.

### ISSUE THREAD

**Issue by tbaederr** [llvm/llvm-project#84871]
For example:
```c++
static_assert(__imag(5.0j / 0) == __builtin_inf());
```
https://godbolt.org/z/59fqKv4Kc

This works without any diagnostics in Clang, but errors out in gcc.

There are actually a few tests in `tests/SemaCXX/complex-folding.cpp` that rely on this behavior, although they aren't as obvious as the example above:
https://github.com/llvm/llvm-project/blob/d02d8df0cd797342f7042440e07133e99ad5e0a2/clang/test/SemaCXX/complex-folding.cpp#L60-L99

In this example:
```c++

constexpr _Complex double D1 = {0.0, 0.0};
constexpr double D2 = __real(__builtin_inf() * D1);
static_assert(D2 == 1);

constexpr double D3 = __builtin_inf() * 0.0;
```
https://godbolt.org/z/dPME9To8d

(which is adapted and simplified from the test above), multiplying `inf` with `0` produces a `nan`, which is properly diagnosed when evaluating the initializer for `D3`, but not at al when evaluating `D2`. The `static_assert` proves that D2 is indeed `nan`.

**Comment by tbaederr** [comment]
@AaronBallman Do you maybe have an opinion here? Should the behavior be changed? Or is this something some projects may rely on?

**Comment by AaronBallman** [comment]
In C, _Complex is morally acceptable in an arithmetic constant expression; C doesn't have a suffix for complex or imaginary numbers so it's technically not possible to have a complex *constant* in standard C, but we support suffixes as an extension so you can have a complex constant.

In C++, `std::complex` has all of its operations marked `constexpr`, so it stands to reason that `_Complex` in C++ would be roughly equivalent.

So I think complex types should follow the usual expectations for use within a constant expression in that it catches UB as being non-constant, but otherwise is fine. (Because we have an extension allowing arbitrary constant expressions in static_assert rather than requiring they be integer constant expressions specifically.) So I think the behavior probably should be changed?

CC @jcranmer-intel for additional opinions.
